### PR TITLE
Feature/pid tuning

### PR DESCRIPTION
**Pull request by Pyrallux**
New pathing_pid.py script uses a similar pathing algorithm to pathing.py, but uses a series to PID loops to smooth and optimize movements while greatly improving drone speed and performance.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,23 +2,25 @@
...
 
 3. Open a command-line interface inside a convenient folder. You will need to navigate to this folder somewhat frequently. Next, clone the ai-racing repository by running the following command. (This requires that you have git installed)
-   git clone https://github.com/MissouriMRR/ai-racing.git
 
-4. Future steps require Python to be installed (The download can be found online). Open the command line and run the following commands to install dependencies.
-   pip install airsimdroneracinglab
-   pip install backports.ssl_match_hostname
+   - `git clone https://github.com/MissouriMRR/ai-racing.git`
+
+4. Future steps require Python to be installed (The download can be found online, multirotor uses python 3.10). Navigate the the home directory of the git repository you just cloned. Open the command line and run the following command to install dependencies.
```
Comment (cjhr95) [inline README.md]: You might need to provide a link to the Python download, some people need the extra help

File: tests/manual_mode.py [hunk 2]
```diff
@@ -94,13 +98,13 @@ def __init__(self, drone_name: str = "drone_1") -> None:
...
-    def repeat_timer(self, callback, period: float) -> None:
+    def repeat_timer(self, callback: Callable[[], None], period: float) -> None:
         """
-        Simple sleep timer.
+        Simple sleep timer used for resetting threads.
 
         Parameters
         ----------
-            callback
+            callback : function
```
Comment (cjhr95) [inline tests/manual_mode.py]: If you're going to use Callable for the type hint, update the docstring as well

File: tests/test_reset.py [hunk 3]
```diff
@@ -89,13 +91,13 @@ def __init__(self, drone_name: str = "drone_1") -> None:
...
-    def repeat_timer(self, callback, period: float) -> None:
+    def repeat_timer(self, callback: Callable[[], None], period: float) -> None:
         """
-        Simple sleep timer.
+        Simple sleep timer used for resetting threads.
 
         Parameters
         ----------
-            callback
+            callback : function
```
Comment (cjhr95) [inline tests/test_reset.py]: Same thing here, update the docstring with the Callable type as well

File: tests/pathing.py [hunk 4]
```diff
@@ -0,0 +1,517 @@
...
+    thread_reset_race : threading.Thread
+        Thread used to reset current race
+    thread_reset_and_reset_race : threading.Thread
+        Thread used to reset airsim client and current race
+    level_name : string
+        Name of current loaded level (see load_level() for options)
+
+    Methods
+    -------
+    repeat_timer(callback, period: float)
```
Comment (cjhr95) [inline tests/pathing.py]: callback parameter needs a type hint. Based on the other callbacks,
`repeat_timer(callback: Callable[[], None], period: float)`

File: tests/pathing.py [hunk 5]
```diff
@@ -0,0 +1,517 @@
...
+        )
+        self.is_thread_active = False
+
+    def repeat_timer(self, callback: Callable[[], None], period: float) -> None:
+        """
+        Simple sleep timer used for resetting threads.
+
+        Parameters
+        ----------
+            callback : function
```
Comment (cjhr95) [inline tests/pathing.py]: Update the type hint

File: tests/pathing.py [hunk 6]
```diff
@@ -0,0 +1,517 @@
...
+        print(time.time(), "called reset")
+        self.airsim_client.reset()
+
+    def reset_race(self) -> None:
+        """Resets current race."""
+        print(time.time(), "called simResetRace")
+        self.airsim_client_2.simResetRace()
+
+    def reset_and_reset_race(self) -> None:
+        """Resets airsim cleint and current race"""
```
Comment (cjhr95) [inline tests/pathing.py]: misspelled client, absolutely cringe

File: tests/pathing.py [hunk 7]
```diff
@@ -0,0 +1,517 @@
...
+        a fixed interval of time.
+
+        Returns
+        -------
+            y_velocity : float
+                Drone's current y velocity (perpendicular left-right velocity) in relation to frame
+        """
+        interval: float = 0.01
+        velocity_deadzone: float = (
+            0.15  # Adjusting this value may result in better/worse performance
```
Comment (cjhr95) [inline tests/pathing.py]: Could you give more description here? Like does increasing the deadzone value improve performance or vice versa?

File: tests/pathing.py [hunk 8]
```diff
@@ -0,0 +1,517 @@
...
+    else:
+        roll_angle = 0
+    print("Roll angle:", roll_angle)
+
+    return roll_angle
+
+
+if __name__ == "__main__":
+    # Sets up race, initializes drone, loads level, and takes off
+    reproducer = ReproduceResetRaceCondition("drone_1")
```
Comment (cjhr95) [inline tests/pathing.py]: Needs a type hint here, not sure what type the reproducer is supposed to  be

File: tests/pathing.py [hunk 9]
```diff
@@ -0,0 +1,517 @@
...
+    reproducer.load_level("Soccer_Field_Easy")  # Level name can be changed  - see load_level() args
+    NUM_GATES = 12  # There are 12 gates on the default level "Soccer_Field_Easy"
+    reproducer.initialize_drone()
+    reproducer.start_race(1)
+    reproducer.takeoff()
+
+    # Gets the number of gates in the active level
+    NUM_GATES = reproducer.get_num_gates()
```
Comment (cjhr95) [inline tests/pathing.py]: Is there a reason why NUM_GATES is initialized with a value if its set a few lines later? Does the reproducer need a gate number to setup?
Also needs a type hint

File: tests/pathing.py [hunk 10]
```diff
@@ -0,0 +1,517 @@
...
+    # Iterates through each gate in the level
+    for next_gate in range(NUM_GATES):
+        INPUT_DURATION: float = 0.1  # Time interval for giving drone commands (in seconds)
+        drone_pitch_angle: float = (
+            0.02  # Currently a constant, should be calculated with a PID loop
+        )
+
+        # Gets next gate position and orientation
+        gate_pose = reproducer.get_gate_pose(next_gate)
+        gate_orientation = airsimdroneracinglab.utils.to_eularian_angles(gate_pose.orientation)
```
Comment (cjhr95) [inline tests/pathing.py]: These fellas need type hints as well

File: tests/pathing.py [hunk 11]
```diff
@@ -0,0 +1,517 @@
...
+
+        # Gives the drone inputs to move it towards the next gate until it arrives
+        distance_to_gate: float = 10000
+        while distance_to_gate > 1:
+            # Gets the drone's current position and orientation
+            drone_pose = reproducer.get_drone_pose()
+            drone_orientation = airsimdroneracinglab.utils.to_eularian_angles(
```
Comment (cjhr95) [inline tests/pathing.py]: Type hints for these goobers as well

File: tests/pathing.py [hunk 12]
```diff
@@ -0,0 +1,517 @@
...
+                drone_pose.orientation
+            )
+            print("Drone Orientation:", drone_orientation)
+            # Generates vector to next gate, then verifies that the vector is a number
+            vector_to_gate = generate_vector(drone_pose.position, gate_pose.position)
```
Comment (cjhr95) [inline tests/pathing.py]: Provide type hint for the vector

File: tests/pathing.py [hunk 13]
```diff
@@ -0,0 +1,517 @@
...
+            if math.isnan(vector_to_gate[0]):
+                time.sleep(INPUT_DURATION)
+                continue
+            # Calculates various angles needed to point the drone at the next gate
+            target_yaw_angle = generate_yaw_angle(vector_to_gate)
+            drone_y_velocity = reproducer.get_drone_y_velocity()
```
Comment (cjhr95) [inline tests/pathing.py]: Type hints? You better believe it

File: tests/pathing_pid.py [hunk 14]
```diff
@@ -0,0 +1,636 @@
...
+    thread_reset_race : threading.Thread
+        Thread used to reset current race
+    thread_reset_and_reset_race : threading.Thread
+        Thread used to reset airsim client and current race
+    level_name : string
+        Name of current loaded level (see load_level() for options)
+
+    Methods
+    -------
+    repeat_timer(callback, period: float)
```
Comment (cjhr95) [inline tests/pathing_pid.py]: Needs `Callable[[], None]` type annotation

File: tests/pathing_pid.py [hunk 15]
```diff
@@ -0,0 +1,636 @@
...
+        Initializes user drone, enabling API control and arming the vehicle.
+        Sets default values for trajectory tracker gains.
+    start_threads()
+        Starts threads if not already active.
+    stop_threads()
+        Stops threads if not already stopped
+    takeoff()
+        Passes low level inputs to the drone to take of rapidly upward.
+        Inputs last exactly 1 second during which the program sleeps before continuing.
+    give_control_stick_inputs(self, roll, pitch, yaw, z, duration)
```
Comment (cjhr95) [inline tests/pathing_pid.py]: Parameters for this function need type annotations

File: tests/pathing_pid.py [hunk 16]
```diff
@@ -0,0 +1,636 @@
...
+        Reads various control stick inputs from command-line interface
+        and passes them to the drone for a given amount of time.
+    get_gate_pose(gate_index)
```
Comment (cjhr95) [inline tests/pathing_pid.py]: gate_index also needs a type annotation

File: tests/pathing_pid.py [hunk 17]
```diff
@@ -0,0 +1,636 @@
...
+        )
+        self.is_thread_active = False
+
+    def repeat_timer(self, callback: Callable[[], None], period: float) -> None:
+        """
+        Simple sleep timer used for resetting threads.
+
+        Parameters
+        ----------
+            callback : function
```
Comment (cjhr95) [inline tests/pathing_pid.py]: Update docstring with Callable type

File: tests/pathing_pid.py [hunk 18]
```diff
@@ -0,0 +1,636 @@
...
+            self.thread_reset_and_reset_race.join()
+            print("Stopped threads.")
+
+    def takeoff(self, orientation: tuple[float, float, float]) -> None:
+        """
+        Passes low level inputs to the drone to take of rapidly upward.
+        Inputs last exactly 1 second during which the program sleeps before continuing.
+
+        Args:
+        -----
```
Comment (cjhr95) [inline tests/pathing_pid.py]:
```suggestion
        Parameters
        ----------
```

File: tests/pathing_pid.py [hunk 19]
```diff
@@ -0,0 +1,636 @@
...
+            drone_orientation : tuple[roll, pitch, yaw]
```
Comment (cjhr95) [inline tests/pathing_pid.py]: The tuple would have parameters of floats, but this works too, I kinda like it

File: tests/pathing_pid.py [hunk 20]
```diff
@@ -0,0 +1,636 @@
...
+            velocity_vector.x_val * math.cos(yaw_angle)
+        )
+
+        y_velocity: float = (velocity_vector.y_val * math.sin(yaw_angle + (math.pi / 2))) + (
+            velocity_vector.x_val * math.sin(yaw_angle)
+        )
+
+        z_velocity: float = -velocity_vector.z_val
+
+        # if y velocity is inside deadzone set it to 0
```
Comment (cjhr95) [inline tests/pathing_pid.py]: Change comment here to reflect the update of all 3 velocity components

File: tests/pathing_pid.py [hunk 21]
```diff
@@ -0,0 +1,636 @@
...
+        """
+        Gets the total number of gates in the active level from the airsim client.
+
+        Returns
+        -------
+            num_gates : int
+                The number of gates present on the active level
+        """
+        gate_list: list[str] = self.airsim_client.simListSceneObjects("Gate.*")
+        num_gates = len(gate_list)
```
Comment (cjhr95) [inline tests/pathing_pid.py]: Silly little type hint

File: tests/pathing_pid.py [hunk 22]
```diff
@@ -0,0 +1,636 @@
...
+    for component in target_vector:
+        distance += math.pow(component, 2)
+    distance = math.sqrt(distance)
+    print(f"Distance to target: {distance}")
+    return distance
+
+
+if __name__ == "__main__":
+    # Sets up race, initializes drone, loads level, and takes off
+    reproducer = ReproduceResetRaceCondition("drone_1")
```
Comment (cjhr95) [inline tests/pathing_pid.py]: Reproducer type hint

File: tests/pathing_pid.py [hunk 23]
```diff
@@ -0,0 +1,636 @@
...
+    # Note that the drone is only optimized for Soccer_Field_Easy and Soccer_Field_Medium
+    reproducer.load_level("Soccer_Field_Easy")  # Level name can be changed - see load_level()
+    reproducer.initialize_drone()
+    reproducer.start_race(1)
+
+    # Gets drone's orientation to take off in the right direction
+    drone_pose: airsimdroneracinglab.Pose = reproducer.get_drone_pose()
+    drone_orientation = airsimdroneracinglab.utils.to_eularian_angles(drone_pose.orientation)
```
Comment (cjhr95) [inline tests/pathing_pid.py]: type hint here

File: tests/pathing_pid.py [hunk 24]
```diff
@@ -0,0 +1,636 @@
...
+    reproducer.takeoff(drone_orientation)
+
+    # Gets the number of gates in the active level
+    NUM_GATES = reproducer.get_num_gates()
```
Comment (cjhr95) [inline tests/pathing_pid.py]:
```suggestion
    NUM_GATES: int = reproducer.get_num_gates()
```

File: tests/pathing_pid.py [hunk 25]
```diff
@@ -0,0 +1,636 @@
...
+    roll_PID: PID = PID(kp=0.14, max_output=(4 * math.pi) / 9)
+    roll_PID.set_target(0)  # Target y (left-right) velocity is 0
+
+    # Iterates through each gate in the level
+    for next_gate in range(NUM_GATES):
+        INPUT_DURATION: float = 0.1  # Time interval for giving drone commands (in seconds)
+
+        # Gets next gate position and orientation
+        gate_pose = reproducer.get_gate_pose(next_gate)
+        gate_orientation = airsimdroneracinglab.utils.to_eularian_angles(gate_pose.orientation)
```
Comment (cjhr95) [inline tests/pathing_pid.py]: Some gate type hints

File: tests/pathing_pid.py [hunk 26]
```diff
@@ -0,0 +1,636 @@
...
+        z_velocity_PID.set_target(gate_pose.position.z_val)  # Target height is middle of next gate
+
+        distance_to_gate: float = 10000
+        # Gives the drone inputs to move it towards the next gate until it arrives
+        while distance_to_gate > 1.1:
```
Comment (cjhr95) [inline tests/pathing_pid.py]: Should 1.1 be a constant?

File: tests/pathing_pid.py [hunk 27]
```diff
@@ -0,0 +1,636 @@
...
+            # Gets the drone's current position and orientation
+            drone_pose = reproducer.get_drone_pose()
+            drone_orientation = airsimdroneracinglab.utils.to_eularian_angles(
```
Comment (cjhr95) [inline tests/pathing_pid.py:604]: drone type hints

File: tests/pathing_pid.py [hunk 28]
```diff
@@ -0,0 +1,636 @@
...
+                drone_pose.orientation
+            )
+            print("Drone Orientation:", drone_orientation)
+            # Generates vector to next gate, then verifies that the vector is a number
+            vector_to_gate = generate_vector(drone_pose.position, gate_pose.position)
```
Comment (cjhr95) [inline tests/pathing_pid.py]: Needs vector type hint

File: tests/pathing_pid.py [hunk 29]
```diff
@@ -0,0 +1,636 @@
...
+            ] = reproducer.get_drone_velocity_components(-drone_orientation[2])
+            drone_x_velocity: float = drone_velocity_components[0]
+            drone_y_velocity: float = drone_velocity_components[1]
+            drone_z_velocity: float = drone_velocity_components[2]
+            # Uses PID loops to calculate and optimize drone inputs
+            drone_z_velocity_input: float = -z_velocity_PID.adjust_output(drone_pose.position.z_val)
+            throttle_PID.set_target(drone_z_velocity_input)
+            drone_throttle_input: float = 0.5938 + throttle_PID.adjust_output(drone_z_velocity)
+            drone_roll_angle = roll_PID.adjust_output(drone_y_velocity)
+            drone_pitch_angle = pitch_PID.adjust_output(drone_x_velocity)
```
Comment (cjhr95) [inline tests/pathing_pid.py]: Float type hints perhaps?

**Review by cjhr95 — CHANGES_REQUESTED** [review]
Mega mondo PR

File: tests/pathing.py (hunk 9 above)
Comment (Pyrallux) [reply tests/pathing.py]: This was old code used before the number of gates was dynamically set based on the number of gates on the current level. The old assignment has been removed and a type hint has been added.

File: tests/pathing.py [hunk 30]
```diff
@@ -480,27 +479,25 @@ def calc_roll_angle(y_velocity: float) -> float:
             0.02  # Currently a constant, should be calculated with a PID loop
         )
 
-        # Gets next gate position and orientation
-        gate_pose = reproducer.get_gate_pose(next_gate)
-        gate_orientation = airsimdroneracinglab.utils.to_eularian_angles(gate_pose.orientation)
+        # Gets next gate position
+        gate_pose: airsimdroneracinglab.Pose = reproducer.get_gate_pose(next_gate)
```
Comment (cjhr95) [inline tests/pathing.py:483]: Is the drone orientation no longer needed?

Comment (Pyrallux) [reply tests/pathing.py:483]: It was originally used in pathing calculations and for debugging, but after breaking many of the calculations into specific functions it is no longer needed.

**Review by cjhr95 — APPROVED** [review]

**Pull request merged**